Foundation/URLSession/http/HTTPURLProtocol.swift

```diff
@@ -189,16 +163,78 @@ fileprivate extension _HTTPURLProtocol {
             self?.completeTask(withError: urlError)
             self?.client?.urlProtocol(self!, didFailWithError: urlError)
         }
-	guard let task = self.task else { fatalError() }
+        guard let task = self.task else { fatalError() }
         easyHandle.timeoutTimer = _TimeoutSource(queue: task.workQueue, milliseconds: timeoutInterval, handler: timeoutHandler)
-
+        
         easyHandle.set(automaticBodyDecompression: true)
         easyHandle.set(requestMethod: request.httpMethod ?? "GET")
         if request.httpMethod == "HEAD" {
             easyHandle.set(noBody: true)
         }
     }
+    
+    /// What action to take
+    override func completionAction(forCompletedRequest request: URLRequest, response: URLResponse) -> _CompletionAction {
+        // Redirect:
+        let httpURLResponse = response as!  HTTPURLResponse
+        if let request = redirectRequest(for: httpURLResponse, fromRequest: request) {
+            return .redirectWithRequest(request)
+        }
+        return .completeTask
+    }
+    
+    override func redirectFor(request: URLRequest) {
+        //TODO: Should keep track of the number of redirects that this
+        // request has gone through and err out once it's too large, i.e.
+        // call into `failWith(errorCode: )` with NSURLErrorHTTPTooManyRedirects
+        guard case .transferCompleted(response: let response, bodyDataDrain: let bodyDataDrain) = self.internalState else {
+            fatalError("Trying to redirect, but the transfer is not complete.")
+        }
+        
+        guard let session = task?.session as? URLSession else { fatalError() }
+        switch session.behaviour(for: task!) {
+        case .taskDelegate(let delegate):
+            // At this point we need to change the internal state to note
+            // that we're waiting for the delegate to call the completion
+            // handler. Then we'll call the delegate callback
+            // (willPerformHTTPRedirection). The task will then switch out of
+            // its internal state once the delegate calls the completion
+            // handler.
+            
+            //TODO: Should the `public response: URLResponse` property be updated
+            // before we call delegate API
+            
+            self.internalState = .waitingForRedirectCompletionHandler(response: response, bodyDataDrain: bodyDataDrain)
+            // We need this ugly cast in order to be able to support `URLSessionTask.init()`
```

Comment:
Is this comment relevant any more?

Reply:
Though I felt it as irrelevant, I have just retained the old comment as I am not sure of the reason behind this comment.